File: postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/ContainerHelper.java
package com.github.aysnc.sql.db.integration;

import com.github.jasync.sql.db.Configuration;
import com.github.jasync.sql.db.Connection;
import com.github.jasync.sql.db.postgresql.PostgreSQLConnection;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testcontainers.containers.PostgreSQLContainer;

import java.util.concurrent.TimeUnit;

/**
 * See run-docker-postresql.sh to run a local instance of postreSql.
 */
public class ContainerHelper {
    protected static final Logger log = LoggerFactory.getLogger(ContainerHelper.class);

    public static PostgreSQLContainer postresql;

    public static Integer getPort() {
        return defaultConfiguration.getPort();
    }

    /**
     * default config is a local instance already running on port 33306 (i.e. a docker postresql)
     */
    public static Configuration defaultConfiguration = new Configuration(
            "postresql_async",
            "localhost",
            33306,
            "root",
            "postresql_async_tests");

    /**
     * config for container.
     */
    private static Configuration rootConfiguration = new Configuration(
            "root",
            "localhost",
            33306,
            "test",
            "postresql_async_tests");


    static {
        try {
            new PostgreSQLConnection(rootConfiguration).connect().get(1, TimeUnit.SECONDS);
            log.info("Using local postresql instance {}", defaultConfiguration);
        } catch (Exception e) {
            // If local instance isn't running, start a docker postresql on random port
            if (postresql == null) {
                postresql = new PostgreSQLContainer("postgresql:9.3")
                        .withDatabaseName("postresql_async_tests")
                        .withPassword("root")
                        .withUsername("postresql_async");
            }
            if (!postresql.isRunning()) {
                postresql.start();
            }
            defaultConfiguration = new Configuration(postresql.getUsername(), "localhost", postresql.getFirstMappedPort(), postresql.getPassword(), postresql.getDatabaseName());
            rootConfiguration = new Configuration("root", "localhost", postresql.getFirstMappedPort(), "test", "postresql_async_tests");
            log.info("Using test container instance {}", defaultConfiguration);
        } finally {
            try {
                Connection connection = new PostgreSQLConnection(rootConfiguration).connect().get(1, TimeUnit.SECONDS);
//        connection.sendQuery("GRANT ALL PRIVILEGES ON *.* TO 'mysql_async'@'%' IDENTIFIED BY 'root' WITH GRANT OPTION;").get(1, TimeUnit.SECONDS);
//        QueryResult r  = connection.sendQuery("select count(*) as cnt  from mysql.user where user = 'mysql_async_nopw';").get(1, TimeUnit.SECONDS);
//        if (r.getRows() != null && r.getRows().size() > 0) {
//          RowData rd = r.getRows().get(0);
//          Boolean exists = ((Long) rd.get(0)) > 0;
//          if (!exists) {
//            connection.sendQuery("CREATE USER 'mysql_async_nopw'@'%'").get(1, TimeUnit.SECONDS);
//          }
//          connection.sendQuery("create table IF NOT EXISTS mysql_async_tests.transaction_test (id varchar(255) not null, primary key (id))").get(1, TimeUnit.SECONDS);
//        }
//        connection.sendQuery("GRANT ALL PRIVILEGES ON *.* TO 'mysql_async_nopw'@'%' WITH GRANT OPTION").get(1, TimeUnit.SECONDS);
            } catch (Exception e) {
                log.error(e.getLocalizedMessage(), e);
            }
        }
    }
}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/ContainerHelper.kt
package com.github.aysnc.sql.db.integration

import com.github.jasync.sql.db.Configuration
import com.github.jasync.sql.db.postgresql.PostgreSQLConnection
import java.util.concurrent.TimeUnit
import mu.KotlinLogging
import org.testcontainers.containers.BindMode
import org.testcontainers.containers.PostgreSQLContainer

private val logger = KotlinLogging.logger {}

private val version = "9.3"

/**
 * See run-docker-postgresql.sh to run a local instance of postgreSql.
 */
object ContainerHelper {
    var postgresql: MyPostgreSQLContainer? = null

    /**
     * default config is a local instance already running on port 15432 (i.e. a docker postgresql)
     */
    var defaultConfiguration = Configuration(
        "postresql_async",
        "localhost",
        15432,
        "root",
        "netty_driver_test"
    )

    init {
        try {
            PostgreSQLConnection(defaultConfiguration).connect().get(1, TimeUnit.SECONDS)
            logger.info("Using local postgresql instance $defaultConfiguration")
        } catch (e: Exception) {
            // If local instance isn't running, start a docker postgresql on random port
            if (postgresql == null) {
                configurePostgres()
            }
            if (!postgresql!!.isRunning()) {
                postgresql!!.start()
            }
            defaultConfiguration = Configuration(
                postgresql!!.getUsername(),
                "localhost",
                postgresql!!.getFirstMappedPort(),
                postgresql!!.getPassword(),
                postgresql!!.getDatabaseName()
            )
            logger.info("PORT is " + defaultConfiguration.port)
            logger.info("Using test container instance {}", defaultConfiguration)
        } finally {
            try {
                val connection = PostgreSQLConnection(defaultConfiguration).connect().get(1, TimeUnit.SECONDS)
                logger.info("got connection " + connection.isConnected())
                connection.sendQuery(
                """
          DROP TYPE IF EXISTS example_mood; CREATE TYPE example_mood AS ENUM ('sad', 'ok', 'happy')
        """
                ).get()
                connection.sendQuery(
                """
          CREATE USER postgres_cleartext WITH PASSWORD 'postgres_cleartext'; GRANT ALL PRIVILEGES ON DATABASE ${defaultConfiguration.database} to postgres_cleartext;
          CREATE USER postgres_md5 WITH PASSWORD 'postgres_md5'; GRANT ALL PRIVILEGES ON DATABASE ${defaultConfiguration.database} to postgres_md5;
          CREATE USER postgres_kerberos WITH PASSWORD 'postgres_kerberos'; GRANT ALL PRIVILEGES ON DATABASE ${defaultConfiguration.database} to postgres_kerberos;
        """
                ).get()
            } catch (e: Exception) {
                logger.error(e.localizedMessage, e)
            }
        }
    }

    private fun configurePostgres() {
        postgresql = MyPostgreSQLContainer()
            .withDatabaseName("netty_driver_test")
            .withPassword("root")
            .withUsername("postresql_async")
            .withClasspathResourceMapping("pg_hba.conf", "/docker-entrypoint-initdb.d/pg_hba.conf", BindMode.READ_WRITE)
            .withClasspathResourceMapping("server.cert.txt", "/docker-entrypoint-initdb.d/server.crt", BindMode.READ_WRITE)
            .withClasspathResourceMapping("server.key.txt", "/docker-entrypoint-initdb.d/server.key", BindMode.READ_WRITE)
            .withClasspathResourceMapping("update-config.sh", "/docker-entrypoint-initdb.d/update-config.sh", BindMode.READ_WRITE)
    }
}

class MyPostgreSQLContainer : PostgreSQLContainer<MyPostgreSQLContainer>("postgres:$version")
